pg_search/src/postgres/customscan/hook.rs

```diff
@@ -802,69 +957,123 @@ unsafe fn replace_windowfuncs_in_query(
 
-            // Get the window target list for this target entry
-            if let Some(window_tl) = window_tls.get(&idx) {
-                let json = serde_json::to_string(window_tl)
-                    .expect("Failed to serialize WindowSpecification");
-
-                // Create a Const node for the JSON string
-                let json_cstring = std::ffi::CString::new(json).expect("Invalid JSON string");
-                let json_text = pg_sys::cstring_to_text(json_cstring.as_ptr());
-                let json_datum = pg_sys::Datum::from(json_text as usize);
-
-                // Create an argument list with the JSON string
-                let mut args = PgList::<pg_sys::Node>::new();
-                let json_const = pg_sys::makeConst(
-                    pg_sys::TEXTOID,
-                    -1,
-                    pg_sys::DEFAULT_COLLATION_OID,
-                    -1,
-                    json_datum,
-                    false, // not null
-                    false, // not passed by value (text is varlena)
-                );
-                args.push(json_const.cast());
-
-                // Create a FuncExpr that calls paradedb.window_agg(json)
-                let funcexpr = pg_sys::makeFuncExpr(
-                    window_agg_procid,
-                    window_tl.singleton_result_type_oid(),
-                    args.into_pg(),
-                    pg_sys::InvalidOid,
-                    pg_sys::InvalidOid,
-                    pg_sys::CoercionForm::COERCE_EXPLICIT_CALL,
-                );
+            // Replace WindowFunc in the expression tree
+            let new_expr = replace_in_node(
+                (*te).expr as *mut pg_sys::Node,
+                window_agg_procid,
+                window_tl,
+            );
 
-                // Replace the WindowFunc with our placeholder FuncExpr
-                (*new_te).expr = funcexpr.cast();
-                new_targetlist.push(new_te);
-                replaced_count += 1;
-            } else {
-                // Still copy the entry but don't replace it
-                new_targetlist.push(te);
-            }
+            (*new_te).expr = new_expr.cast();
+            new_targetlist.push(new_te);
         } else {
-            // For non-WindowFunc entries, just make a flat copy
+            // No window function in this entry - just copy it
             let copied_te = pg_sys::flatCopyTargetEntry(te);
             new_targetlist.push(copied_te);
         }
     }
 
     (*parse).targetList = new_targetlist.into_pg();
 }
+
+// Helper function to recursively replace WindowFunc nodes in an expression
+//
+// Note: This follows a similar recursive pattern to replace_window_agg_with_const() in mod.rs,
+// but operates at a different stage:
+// - This function: Planning stage - replaces WindowFunc → window_agg() placeholder
+// - That function: Execution stage - replaces window_agg() → Const placeholder for value injection
```

Comment:
Please add a reference to https://github.com/paradedb/paradedb/issues/3455 here.